### PR TITLE
Search for frontend in redis using trailing slashes

### PR DESCRIPTION
**Pull request by kippandrew**
CC @aronatkins @CBarraford

### CODE REVIEW

File: lua/src/lib/spec/library_spec.lua [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    assert.are.same(library.strip('//derp//', '/'), 'derp')
+    return assert.are.same(library.strip('//derp// ', '/'), 'derp// ')
+  end)
+end)
+return describe("trim function", function()
+  return it("should remove leading and trailing whitespace", function()
+    assert.are.same(library.trim('  derp '), 'derp')
+    assert.are.same(library.trim('derp'), 'derp')
+    assert.are.same(library.trim('derp\n'), 'derp')
+    return assert.are.same(library.trim('\nderp\n'), 'derp')
```
Comment (aronatkins) [inline lua/src/lib/spec/library_spec.lua]: nit: tab stop?

File: lua/src/lib/redis.lua [hunk 2]
```diff
@@ -538,9 +539,11 @@ M.fetch_frontend = function(self, max_path_length)
     if not (v == nil or v == '') then
       if count < (max_path_length) then
         count = count + 1
+        v = library.strip(v, '/')
```
Comment (aronatkins) [inline lua/src/lib/redis.lua]: Should this be a `rstrip` not `strip`?

Comment (aronatkins) [inline lua/src/lib/redis.lua]: never mind. i see where `p` is getting a slash re-added.

**Comment by aronatkins** [comment]
:hamburger:

**Comment by CBarraford** [comment]
So i'm a little concerned that we're hardcoding a lucid specific behavior into an open source project. I don't think its reasonable that other people would want this same behavior.

**Comment by CBarraford** [comment]
I'd like to keep redx agnostic to what its proxying to. Keep it a general use application, not something for lucid specifically. In order to do that, we should be heading towards staying true to RFCs around urls rather than moving away to standards.

**Comment by prologic** [comment]
:+1: to keeping this agnostic. As the developer of [autodock](https://github.com/prologic/autodock) and [autodock-hipache](https://github.com/prologic/autodock-hipache) I'd love to explore the possibility of writing a new alternative "load balancer" plugin; [autodock-redx](TBD) :)

**Comment by kippandrew** [comment]
The goal here is not to be coupled to rstudio's internal code base, but rather remove the ambiguity in how redx resolves frontends:

Because its not possible to do any pattern matching when fetching frontends from redis, the frontend has to be an exact match. This means that assuming a request for: `www.example.com/foo/`, redx will not find a matching frontend if the configured frontend URL is `www.example.com/foo`. But a request for `www.example.com/foo/bar` will match. 

To me this means is that the actual pattern we are trying to match is `www.example.com/foo/*`. Given this, it makes most sense to represent this internally using a URI that ends in slash, which is to clarify what URLs should be matched.

**Comment by prologic** [comment]
Q: How many lb's actually support matching on Path? I'm not sure that hipache does (*I could be wrong*); I think haproxy supports this?

**Comment by kippandrew** [comment]
I know nginx and and haproxy can do this (or any high end LB appliance: F5, Netscaler, etc). Not sure about hipache though.

If you're not looking to match on path with redx, this change probably doens't matter.

**Comment by prologic** [comment]
I also only because of my previous comment to this; I should start considering addthis this as a feature to autodock; but I'd also like to support redx as well in the near future.

File: lua/src/lib/spec/library_spec.lua (hunk 1 above)
Comment (kippandrew) [inline lua/src/lib/spec/library_spec.lua]: This is auto generate code.

Comment (aronatkins) [inline lua/src/lib/spec/library_spec.lua]: I know. I meant. Should you test `'\t'` as well. Sorry if I read the Lua code first. Alpha sort. :fallen_leaf:

Comment (kippandrew) [inline lua/src/lib/spec/library_spec.lua]: Ah. Thats fine. I was super confused. I though you were complaining about the indentation level of the last line. :tulip:

**Comment by CBarraford** [comment]
@prologic can you rephrase your last comment? Not sure i understand what you're saying.

**Comment by CBarraford** [comment]
@kippandrew Sorry, this merged quicker than i thought it was going to, please update the tests to include trailing slashes https://github.com/rstudio/redx/blob/master/spec/api_spec.moon

File: src/lib/spec/library_spec.moon [hunk 3]
```diff
@@ -0,0 +1,22 @@
+library = require 'library'  -- from redx
```
Comment (CBarraford) [inline src/lib/spec/library_spec.moon]: This file is in the wrong directory, should be in `/spec` with all the other tests

File: docs/api.md [hunk 4]
```diff
@@ -30,7 +30,7 @@ curl -X POST localhost:8081/frontends/myhost.com%2Ftest/mybackend
```
Comment (CBarraford) [inline docs/api.md]: if trailing slashes are now required, this POST/PUT request (line 28) is now broken, thus the api is broken

**Comment by CBarraford** [comment]
@kippandrew also add tests to `main_spec.moon` showing that using trailing slashes and not still works ok.
